### ISSUE TITLE
Cloud Run ESPv2 timed out even though Cloud function returned with 200 response

### ISSUE THREAD

**Issue by mejuhi** [GoogleCloudPlatform/esp-v2#503]
We have Architecture where we have configured Cloud Run and Cloud Functions to connect using Cloud Endpoint. We have used following [documentation](https://cloud.google.com/endpoints/docs/openapi/get-started-cloud-functions). In which the ESP image is running on Cloud run

Base image used by cloud run is gcr.io/endpoints-release/endpoints-runtime-serverless:2.24.0 which gets build using the following [script](https://github.com/GoogleCloudPlatform/esp-v2/blob/40caa16ea9b4231ab9272bce2c9dfcec95dee3e5/docker/serverless/gcloud_build_image)

Cloud run (ESPv2) is running with following configuration:

CPU allocated: 2
Memory allocated: 2048
Concurrency: 80
Request timeout: 800 seconds
Auto-scaling: Min instances: 5
Auto-scaling: Max instances: 1,000

Additionally we have added following arguments in the base docker image:
ENV ESPv2_ARGS ^++^--cors_preset=basic++--cors_allow_method="GET,PUT,POST"++--backend_retry_num=2

The Cloud Function is API Business Layer. Cloud function was able to execute successfully and returned response with 200 HTTP status code, but somehow Cloud Run ESP timed out and then Cloud endpoint timed out with HTTP status code 504. 

Attaching screenshot:
![image](https://user-images.githubusercontent.com/45816852/116202142-e4f22580-a757-11eb-8bde-f33a0327eb91.png)

**Comment by qiwzhang** [comment]
Looking at the log from your screenshot.  It seems that 504 is from time-out.  The last two lines have  800s and 3600s respectively.   

A couple questions:
* Does 504 happens to all requests or only to some?
* If only for some, do these requests have long latency to response?

**Comment by mejuhi** [comment]
1. It has happened only few times
2. No, as you can see the third last line in the screenshot, cloud function was able to complete its execution in 1043 ms with 200, but still cloud run running ESPv2 timed out for some reason.

**Comment by qiwzhang** [comment]
backend default timeout is 15s.  It seems that your backend responded within a couple seconds so it should not be the default timeout issue.
How about the data size,  your screen shot blocked out the data size,   what are the request and response data size?

**Comment by mejuhi** [comment]
Http request size i can see from the log of Cloud Endpoint is around 2846 B (attached screenshot below), and response size should be around 1274 B (from the cloud function)

![image](https://user-images.githubusercontent.com/45816852/116293125-24e6f600-a7b4-11eb-934b-3731fc17cc3a.png)

**Comment by qiwzhang** [comment]
From the Cloud Endpoint log,  it has response_code_detail: as upstream_response_timeout.  It for sure implies that ESPv2 did not receive the response within 15s.  or whatever timeout you specified in `deadline` inside the `x-google-backend` in your OpenApi spec.   The latency is 3600s, you may have change the timeout to 3600s. 

It seems that  ESPv2 has send the request to your Cloud Function backend, Cloud Function logs confirm that.  but ESPv2 never receive its response, even though your Cloud Function log indicates that the call is completed.  

Could you dig more on the Cloud Function logs to see where the response goes?   Since this only happen intermittently, I will not suspect it is a config problem,  it is more like a network problem.

**Comment by mejuhi** [comment]
We have defined the deadline timeout already in the OpenApi spec as 3600s, attaching the specific definition of this integration. It can also be confirmed from the last log in the screenshot that it waited for 3600s.

I did try to look out for all the logs having the trace id, but only logs which I can see are attached in the first screenshot. Not sure if there are some additional logs which can be accessed by internal GCP team which can help to understand where exactly the response got lost? 
```  
"/<integration>":
    post:
      description: "API for <integration>"
      operationId: "<integration>"
      x-google-backend:
        address: https://<REGION>-<PROJECT-ID>.cloudfunctions.net/<integration> # Backend Cloud function URL
        deadline: 3600.0
      # Defines Authentication mechanism to use, Following mentions to use API KEYS
      security:
      - api_key: []
      # MIME Types expected as request and response        
      produces:
      - "application/xml"     
      consumes:
      - "application/xml"  
      responses:
        200:
          description: OK    
        404:
          description: Not Found
        500:
          description: Internal Service Error  
```

**Comment by qiwzhang** [comment]
Do you have any support channel for Google Cloud?  If you do, please contact them

**Comment by mejuhi** [comment]
I can create GCP support ticket in GCP project and check if they are able to find anything more. I can comment back on the findings. 
Just curious to know, is there possible workaround in your mind which can prevent these situations from happening on cloud run ESP?

**Comment by qiwzhang** [comment]
I don't the root cause of the issue so it is very hard for me to recommend a work around.

**Comment by mejuhi** [comment]
Yes, I agree. I will create support ticket for it and let you know about the findings. Thankyou! :)

**Comment by mejuhi** [comment]
I had a question, In scenario like this, we http status code we get from Cloud run ESP is 504. 

Workaround for this problem we were thinking to do retry from Cloud run ESP side for requests with 504 Http status code  like you have provided the retry solution for 503 Http status code errors for this [problem ](https://github.com/GoogleCloudPlatform/esp-v2/issues/430)

We wanted to know if we can have some additional configuration in the docker image of Cloud run ESP to perform retry on both status code (503 & 504) and will this type of retry via envoy even work in this situation

**Comment by qiwzhang** [comment]
Hi @TAOXUY what "retry_on" flags @mejuhi can use for (503 or 504)?   I think we already have default "retry_on" for connection_reset

BTW, you may gave to reduce yout deadline.  Now it is 1 hours, you don't want to retry for another hour.

**Comment by qiwzhang** [comment]
Well,  it is upstream timeout,  I don't think "retry_on" will help.    ESP envoy did not receive any response from the upstream.

**Comment by TAOXUY** [comment]
The default `backend_retry_ons=reset,connect-failure,refused-stream` is for the connection issue. The http error code for that is 503 for the connection issue based on our [test](https://github.com/TAOXUY/esp-v2/blob/c2462a6af11d23ee43900ac2f3f617bbc8738581/tests/integration_test/backend_retry_test.go#L85). Not sure if it covers your timeout case(504). You can try to add flag `backend_retry_ons=reset,connect-failure,refused-stream,gateway-error`  to do retry on 504.

- https://github.com/TAOXUY/esp-v2/blob/c2462a6af11d23ee43900ac2f3f617bbc8738581/docker/generic/start_proxy.py#L412-L430
- https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/router_filter#x-envoy-retry-on

**Comment by mejuhi** [comment]
Okay, Yes, I will add the flag in the docker image. but do you think it will work for this particular scenario, since @qiwzhang  mentioned it might not work?
Is there any way to recreate and test it out, since this error occurs randomly, not sure how to replicate such behavior

**Comment by qiwzhang** [comment]
if occurs randomly, it will be difficult to reproduce it in other setup. 

How big is your regular traffic,  if your traffic is less then 10 request per second,  I will suggest you turn on `--enable_debug` flag in ESPv2,  envoy will generate more debug log.  It may provide more clues.    If the traffic is more than 10 qps,  it will `--enable_debug`` will generate too much log.

**Comment by mejuhi** [comment]
We do have traffic more than 10 qps in our prod, and we have only seen this error in the prod environment.

**Comment by qiwzhang** [comment]
With such heavy traffic,  I have suspicious of you may mismatch the requests with 504 at ESPv2 and the ones at your backend cloud function.  General assumption is: ESPv2 creates a tcp connection to CF,  then trigger a call in CF,  that same connection should deliver the response back.  Normally,  you will not suspect any network issue of sending request is working, but receiving response is not.  But from Envoy "response_code_detail", it indicates it is upstream timeout,  it means,  tcp connection still alive, but response is not received.   So I suspected that the log entries from the Cloud Function may not from the one with 504 at ESPv2.

**Comment by qiwzhang** [comment]
In another word,  I suspect that your CF indeed was taking too long (more than 1 hour) to process the request.

**Comment by mejuhi** [comment]
Yes, but to avoid such confusion we check the logs by filtering out based on trace.  From what I understand trace id gets generated(on cloud endpoint/ cloud run level) and passed  down to cloud function as a header `x-cloud-trace-context`, all the logging on the cloud function makes use of that trace id and hence we can see end to end flow of the message.

We did suspect and hence we did double check the timings and trace id on all the logs. Moreover there was no other requests from our cloud function which failed or took more than 1 sec to respond (we checked all the requests for that day). Also to mention, when this error occurred it was not peak time and hence traffic was also quite low

**Comment by TAOXUY** [comment]
Hi @mejuhi, I think I reproduced the error locally, which is the `upstream_response_timeout`. It won't be retried[1] by Envoy if we only setting `gateway-error` as the deadline(3600 as you set) has been exhausted. Instead, I find if we set  `per_try_timeout`[2] in the same time, it can do retry under this case. However, we haven't exposed this option so far and I can add that as flag. Just let you know the change may take several weeks to be released so I would recommend you go to network/CF channel to root out the issue if possible.

[[1](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/router_filter#x-envoy-retry-on)]*: NOTE: Envoy will not retry when a request exceeds x-envoy-upstream-rq-timeout-ms (resulting in a 504 error code). Use x-envoy-upstream-rq-per-try-timeout-ms if you want to retry when individual attempts take too long. x-envoy-upstream-rq-timeout-ms is an outer time limit for a request, including any retries that take place.*
[[2](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/router_filter#x-envoy-retry-on)]: *Specifies a non-zero upstream timeout per retry attempt. This parameter is optional. The same conditions documented for x-envoy-upstream-rq-per-try-timeout-ms apply.*

**Comment by jugup** [comment]
I think we can wait for this feature for few weeks and then add those parameters for our deployment since this error happen randomly very few times.
Thankyou for looking into it and providing workaround so quickly!

**Comment by mejuhi** [comment]
I think we can wait for this feature for few weeks and then add those parameters for our deployment since this error happen randomly very few times.
Thankyou for looking into it and providing workaround so quickly!

**Comment by TAOXUY** [comment]
The fix should be GA in our next version 2.26.0.
